jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java

```diff
@@ -0,0 +1,465 @@
+        {
+            long encoded = state.get();
+            int permits = AtomicBiInteger.getHi(encoded);
+            if (permits > 0)
+            {
+                if (!state.compareAndSet(encoded, permits - 1, 0))
+                    continue;
+                return forward(request, response, callback);
+            }
+            else
+            {
+                boolean expired = request.getAttribute(EXPIRED_ATTRIBUTE) != null;
+                if (expired)
+                {
+                    // Do not suspend again an attempt to handle the expired request/response.
+                    response.setStatus(HttpStatus.SERVICE_UNAVAILABLE_503);
+                    if (response.isCommitted())
+                        callback.failed(new IllegalStateException("Response already committed"));
+                    else
+                        response.write(true, null, callback);
+                }
+                else
+                {
+                    // Avoid this race condition:
+                    // T1 in handle() may find no permits, so it will suspend the request.
+                    // T2 in resume() finds no suspended requests and increments the permits.
+                    // T1 suspends the request, which will remain suspended despite permits are available.
+                    // Below, the suspended request count is only incremented if there are no permits.
+                    // See correspondent code in resume().
+                    int suspended = AtomicBiInteger.getLo(encoded);
+                    if (!state.compareAndSet(encoded, 0, suspended + 1))
+                        continue;
+                    suspend(request, response, callback);
+                }
+                return true;
+            }
+        }
+    }
+
+    /**
+     * <p>Returns the priority of the given suspended request,
+     * a value greater than or equal to {@code 0}.</p>
+     * <p>Priority {@code 0} is the lowest priority.</p>
+     * <p>The set of returned priorities should be stable over
+     * time, typically constrained in the range {@code 0-10}.</p>
+     *
+     * @param request the suspended request to compute the priority for
+     * @return the priority of the given suspended request, a value {@code >= 0}
+     */
+    protected int getPriority(Request request)
+    {
+        return 0;
+    }
+
+    /**
+     * <p>Fails the given suspended request/response with the given error code and failure.</p>
+     * <p>This method is called only for suspended requests, in case of timeout while suspended,
+     * or in case of failure when trying to handle a resumed request.</p>
+     *
+     * @param request the request to fail
+     * @param response the response to fail
+     * @param callback the callback to complete
+     * @param status the failure status code
+     * @param failure the failure
+     */
+    protected void fail(Request request, Response response, Callback callback, int status, Throwable failure)
```

Comment:
Since this is API, maybe rename to a more explicit name like `failSuspendedRequest()`?